### PR TITLE
ethcore: fix ancient block sync

### LINKED ISSUE #7008: Ancient block sync stalls
> I'm running:
>
> - **Parity version**: Parity/v1.8.2-beta-1b6588c-20171025/x86_64-linux-gnu/rustc1.21.0
> - **Operating system**: Linux
> - **And installed**: sudo apt-get install ./parity_1.8.2_amd64.deb

configuration detail: chain is simlinked to another drive
```
human@EtherBox:~$ ls -lisa ./.local/share/io.parity.ethereum/chains
524382 0 lrwxrwxrwx 1 human human 28 Sep  3 21:25 ./.local/share/io.parity.ethereum/chains -> /media/human/CHAIN_DB/chains
```

After warp sync, I'm unable to get all missing blocks fetched. After new start Parity forgets the already fetched blocks and start downloading them again and again.

Here is a sample of 3 parity runs

image

 
Here is the log file from above (as text file)
[parity-3x-sync.txt](https://github.com/paritytech/parity/files/1456141/parity-3x-sync.txt)

---

### LINKED ISSUE #9300: Background verifier sometimes goes back
> I'm running:
>
> - **Which Parity version?**: 1.11.8
> - **Which operating system?**: Linux
> - **How installed?**: via binaries
> - **Are you fully synchronized?**: no
> - **Which network are you connected to?**: core, sokol
> - **Did you try to restart the node?**: yes

We run a POA-based networks (sokol and core). After upgrade to 1.11.8 I'm getting hard time syncing with the network using warp-sync.
After all snapshots were downloaded, a verifier in the background starts processing blocks (if I understand correctly), here's a piece of log: 
```
2018-08-07 20:21:42    #1320360   15/25 peers   7 MiB chain 87 MiB db 0 bytes queue 5 MiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-08-07 20:21:45  Imported #4050210 0x6620…f959 (2 txs, 0.25 Mgas, 16 ms, 0.85 KiB)
2018-08-07 20:21:51  Imported #4050211 0x7253…604b (1 txs, 0.02 Mgas, 0 ms, 0.67 KiB)
2018-08-07 20:21:55  Imported #4050212 0x3af3…1b8a (0 txs, 0.00 Mgas, 1 ms, 0.57 KiB)
2018-08-07 20:22:01  Imported #4050213 0xc563…69f7 (1 txs, 0.23 Mgas, 3 ms, 0.75 KiB)
2018-08-07 20:22:05  Imported #4050214 0x3187…d824 (0 txs, 0.00 Mgas, 0 ms, 0.57 KiB)
2018-08-07 20:22:10  Imported #4050215 0x688d…38f4 (1 txs, 0.23 Mgas, 4 ms, 0.75 KiB)
2018-08-07 20:22:12    #1345587   15/25 peers   6 MiB chain 87 MiB db 0 bytes queue 5 MiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-08-07 20:22:15  Imported #4050216 0xef42…9d84 (0 txs, 0.00 Mgas, 0 ms, 0.57 KiB)
2018-08-07 20:22:20  Imported #4050217 0x9754…49d1 (2 txs, 0.40 Mgas, 4 ms, 0.99 KiB)
2018-08-07 20:22:25  Imported #4050218 0x5528…1323 (0 txs, 0.00 Mgas, 1 ms, 0.57 KiB)
2018-08-07 20:22:30  Imported #4050219 0x30e4…589e (1 txs, 0.23 Mgas, 2 ms, 0.75 KiB)
2018-08-07 20:22:41  Imported #4050221 0x3dbc…4ce5 (1 txs, 0.08 Mgas, 1 ms, 0.80 KiB) + another 1 block(s) containing 1 tx(s)
2018-08-07 20:22:46  Imported #4050222 0x28b2…f2fa (1 txs, 0.07 Mgas, 5 ms, 0.80 KiB)
2018-08-07 20:22:50    #1375534   15/25 peers   7 MiB chain 87 MiB db 0 bytes queue 85 KiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-08-07 20:22:51  Imported #4050223 0x57a7…9be5 (0 txs, 0.00 Mgas, 4 ms, 0.57 KiB)
2018-08-07 20:22:55  Imported #4050224 0x6b50…3828 (0 txs, 0.00 Mgas, 8 ms, 0.57 KiB)
2018-08-07 20:23:00  Imported #4050225 0x8898…a33b (1 txs, 0.18 Mgas, 8 ms, 0.71 KiB)
2018-08-07 20:23:05  Imported #4050226 0xca15…f9eb (0 txs, 0.00 Mgas, 10 ms, 0.57 KiB)
2018-08-07 20:23:10  Imported #4050227 0x1b17…91f5 (2 txs, 0.40 Mgas, 9 ms, 0.99 KiB)
2018-08-07 20:23:15  Imported #4050228 0x6726…0861 (2 txs, 0.24 Mgas, 4 ms, 1.04 KiB)
2018-08-07 20:23:20  Imported #4050229 0x466d…f469 (3 txs, 0.69 Mgas, 14 ms, 1.10 KiB)
2018-08-07 20:23:22       #8636   15/25 peers   6 MiB chain 87 MiB db 0 bytes queue 1 MiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-08-07 20:23:26  Imported #4050230 0x4092…dace (4 txs, 0.79 Mgas, 10 ms, 1.41 KiB)
2018-08-07 20:23:32  Imported #4050231 0xd356…9b8f (1 txs, 0.18 Mgas, 6 ms, 0.71 KiB)
2018-08-07 20:23:35  Imported #4050232 0xbd55…a91e (0 txs, 0.00 Mgas, 0 ms, 0.57 KiB)
2018-08-07 20:23:41  Imported #4050233 0xd0de…d6d8 (6 txs, 0.84 Mgas, 4 ms, 1.62 KiB)
2018-08-07 20:23:46  Imported #4050234 0x6b69…b6c5 (1 txs, 0.07 Mgas, 3 ms, 0.80 KiB)
2018-08-07 20:23:51  Imported #4050235 0x51e7…665f (0 txs, 0.00 Mgas, 7 ms, 0.57 KiB)
2018-08-07 20:23:56  Imported #4050236 0x20c4…db14 (2 txs, 0.46 Mgas, 22 ms, 0.93 KiB)
2018-08-07 20:23:57      #13716   15/25 peers   6 MiB chain 87 MiB db 0 bytes queue 1 MiB sync  RPC:  0 conn,  0 req/s,   0 µs
2018-08-07 20:24:00  Imported #4050237 0x4672…a188 (0 txs, 0.00 Mgas, 8 ms, 0.57 KiB)
2018-08-07 20:24:06  Imported #4050238 0x9e69…9455 (3 txs, 0.22 Mgas, 12 ms, 0.93 KiB)
2018-08-07 20:24:10  Imported #4050239 0x884a…2481 (2 txs, 0.04 Mgas, 2 ms, 0.78 KiB)
2018-08-07 20:24:15  Imported #4050240 0x63cc…ab77 (5 txs, 0.51 Mgas, 24 ms, 1.43 KiB)
2018-08-07 20:24:20  Imported #4050241 0x86a3…d6ba (3 txs, 0.21 Mgas, 4 ms, 1.02 KiB)
2018-08-07 20:24:25  Imported #4050242 0xa90c…aa55 (3 txs, 0.27 Mgas, 3 ms, 0.96 KiB)
2018-08-07 20:24:31  Imported #4050243 0x51ee…3def (2 txs, 0.04 Mgas, 3 ms, 0.78 KiB)
```
As you can see it goes `1345587 -> 1375534` and then drops to `8636`. This can happen multiple times, so it's hard to get the verification process to complete. It happens on different block numbers

To test it yourself, please use our [spec.json](https://raw.githubusercontent.com/poanetwork/poa-chain-spec/core/spec.json) and [list of reserved peers](https://raw.githubusercontent.com/poanetwork/poa-chain-spec/core/bootnodes.txt)
```
$ parity --chain spec.json --reserved-peers bootnodes.txt
```

### PR DESCRIPTION
**Pull request by andresilva**
Fixes #9300 and fixes #7008. Merge after #9381.

### CODE REVIEW

File: ethcore/sync/src/block_sync.rs [hunk 1]
```diff
@@ -465,9 +465,11 @@ impl BlockDownloader {
 	/// Checks if there are blocks fully downloaded that can be imported into the blockchain and does the import.
 	pub fn collect_blocks(&mut self, io: &mut SyncIo, allow_out_of_order: bool) -> Result<(), BlockDownloaderImportError> {
 		let mut bad = false;
+		let mut reset = false;
 		let mut imported = HashSet::new();
 		let blocks = self.blocks.drain();
 		let count = blocks.len();
+
 		for block_and_receipts in blocks {
```
Comment (tomusdrw) [inline ethcore/sync/src/block_sync.rs]: Might be worth to turn this loop into a closure and return an `enum Importer { Reset, Bad, Imported(HashSet) }` instead of setting the flags.

That said, refactoring can be left to a separate PR.

File: ethcore/sync/src/block_sync.rs [hunk 2]
```diff
@@ -523,15 +533,24 @@ impl BlockDownloader {
 		trace!(target: "sync", "Imported {} of {}", imported.len(), count);
 		self.imported_this_round = Some(self.imported_this_round.unwrap_or(0) + imported.len());
 
-		if bad {
-			return Err(BlockDownloaderImportError::Invalid);
-		}
-
 		if self.blocks.is_empty() {
```
Comment (tomusdrw) [inline ethcore/sync/src/block_sync.rs]: Is that correct? Seems that if we fail on the last block we are not going to return an error?

**Comment by andresilva** [comment]
@tomusdrw I didn't notice any changes on regular sync, seemed to be working fine. But it seems like this PR doesn't fix the original issue completely (https://github.com/paritytech/parity-ethereum/issues/9300#issuecomment-416041318). So I'm going to investigate this a bit more and I'll put this `onice` in the meantime.

**Comment by 5chdn** [comment]
What's the status?

**Comment by ascjones** [comment]
@5chdn PR for a fix in #9531. Just waiting for mainnet sync to confirm.

**Comment by andresilva** [comment]
After discussing with @ascjones and looking into the sync code a bit more, we've concluded that this PR won't fix the issue properly. Closing and hopefully it is fixed in #9531 (will have a look at it tomorrow :slightly_smiling_face:).
